src/oop/programs_with_multiple_classes/hard/q4/MainTest.java

```diff
@@ -0,0 +1,150 @@
+package oop.programs_with_multiple_classes.hard.q4;
+
+import global.ObjectTest;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import java.time.Year;
+import java.util.stream.Stream;
+
+import static global.tools.CustomAssertions._assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class MainTest {
+    // Java
+
+    private ObjectTest book;
+    private ObjectTest bookcase;
+
+    private final String bookLc = "book";
+    private final String bookUc = "Book";
+    private final String bookcaseLc = "bookcase";
+    private final String bookcaseUc = "Bookcase";
+    private final String year = "year";
+    private final String type = "type";
+    private final String sizeCapacity = "sizeCapacity";
+    private final String b1 = "b1";
+
+    private final String howManyYears = "howManyYears";
+
+
+    @BeforeEach
+    public void Setup() {
+        String bookClassString = "oop.programs_with_multiple_classes.hard.q4." + bookUc;
+        String bookcaseClassString = "oop.programs_with_multiple_classes.hard.q4." + bookcaseUc;
+        book = new ObjectTest(bookClassString);
+        bookcase = new ObjectTest(bookcaseClassString);
+    }
+
+    /**
+     * Book tests
+     **/
+    @Test
+    public void bookClassHasCorrectFields() {
+        String missingFieldMessage = "Your " + bookLc + " class is missing a required field.";
+        String incorrectFieldModifierMessage = "A field in your " + bookLc + " class has an the wrong visibility modifier.";
+        assertTrue(book.hasField(year, int.class), missingFieldMessage);
```

Comment:
You can also use the new method I added which allows you to check if the attribute exists before checking if it has the correct type. This way we can give two error messages to the student. One that mentions the attribute could not be found and to check it was added and spelt correctly and another to check that the attribute has the correct type.